website/database.py

```diff
@@ -45,7 +45,8 @@
 # - tags_id (str): id of tags that describe this adventure.
 ADVENTURES = dynamo.Table(storage, "adventures", "id", indexes=[dynamo.Index("creator"), dynamo.Index("public")])
 INVITATIONS = dynamo.Table(
-    storage, "class_invitations", partition_key="username", indexes=[dynamo.Index("class_id")]
+    storage, "class_invitations", partition_key="username#class_id",
```

Comment:
This is a good idea! Unfortunately, DynamoDB does not allow changing the key of an existing table. If you want to do this, you'll have to create a new table (`invitations2` or whatever).

You then have a choice:

* We fully switch over to the new table and effectively invalidate all existing invitations
* We are a more complicated multi-step deployment if we want to make sure that old invitations still work while we are cutting over to the new table

If we do the multi-step deployment to keep all invitations alive:

* Create a new table with the desired keys (`invitations2` or whatever)
* Update the code to **write** to `invitations2`, but still **read** from both `invitations` and `invitations2`.
* Deploy
* Copy over all old invitations from `invitations` to `invitations2` and empty the old table, or wait a while until all old invitations have expired. 
* Remove all references to `invitations` from the code
* Deploy
* Delete the `invitations` table

By the way, a partition key of `username#class_id` with 2 indexes will totally work, but you can also do it like this:

```
partition_key: username
sort_key: class_id
```

If you do that, you already have the index on `username` (it's just the main table) and you only need an additional index on `class_id`.

Reply:
Hi both! 

There are currently no open invitations in the database, so I think we can save ourselves the hassle and migrate directly. I suspect invitations for students are not so often used, most teachers add kids to their class directly, or they are accepted immediately and then deleted from the table.

Reply:
Yep, confirmed, there are `0` rows in the table right now. Good shout @hasan-sh. We can just make a new table in Dynamo and change the table name that the `Table` variable points to, but we don't have to add a new variable if we're only going to do a one-step deployment. I can do that immediately after we settle on a table name and a key schema. 

## Name

Currently the table is called `class_invitations`... how about `invitations` ?

## Key schema

> Does this mean we could mitigate multi-step deployment and still use the same table?

Unfortunately, no. Adding a sort key is still changing the key schema, so we're still going to have to make a new table.

Using the table plus one index, as opposed to the table plus two indexes, just saves you a copy of the data.